Added criterion 'window_type'

Airblader opened the pull request:
~~Don't merge this yet, handling updates to the window type is still missing.~~

fixes #1658

Airblader committed:
Implement new criterion 'window_type = normal|dialog|utility|toolbar|splash|menu|dropdown_menu|popup_menu|tooltip'

fixes #1658

Airblader committed:
Added testcase for criterion 'window_type'.

Airblader committed:
Handle changes to _NET_WM_WINDOW_TYPE after the window has been managed.

Airblader committed:
Added testcase for changing _NET_WM_WINDOW_TYPE after the window is already managed.

Airblader committed:
Added testcase for 'window_type' criterion when used as a command.

Airblader commented:
OK, I've added handling changes to the window type. An interesting question, but a question for a different PR, is whether we should float windows if their type changes to `_NET_WM_WINDOW_TYPE_DIALOG` and the like.

Airblader commented:
I will put all the unrelated cleanups in a separate commit.

Airblader committed:
Small cleanups:
* Implement criterion specs just like it is done in the spec for the config
* Declare variables in test case early so the 'my' keyword can be dropped in the actual test cases

Airblader committed:
Cleanup: delete comments that were accidentally copy-pasted and are incorrect for these test cases

Airblader commented:
@acrisci I've incorporated all the changes and extracted the cleanups into separate commits. I think the only open point was why I didn't use the atom in the match definition. The answer was that interning of the atoms happens after the config has been parsed/loaded, so it doesn't work unless we intern the atoms before we even parse the config (for the `i3 -C` case), but I'm not sure we want to do that.

Airblader commented on include/data.h:
```diff
@@ -409,6 +412,18 @@ struct Match {
+        WT_DROPDOWN_MENU = 6,
+        WT_POPUP_MENU = 7,
+        WT_TOOLTIP = 8
+    } window_type;
```
@acrisci asked why we don't just use the atom here and introduce an enum instead.

Answer: interning of the atoms happens only after the config is loaded, which means `for_window` directives wouldn't be parsed correctly. The only solution would be interning the atoms before even checking the configuration (to not break `i3 -C`), but that doesn't seem feasible to me.

acrisci replied:
I don't think it's such a big deal to get the atoms before we parse the config if we get to avoid this ugly enum. As far as I can tell, it's just a matter of moving some code around in `main.c`.

Airblader replied:
Yes, we just have to move that. However, we have to move it all the way up to before the config is *checked* and currently we don't really do anything X related at that point, so it's sort of a paradigm shift. It certainly would make the rest of this PR look a whole lot better.

acrisci replied:
I'm talking about getting the atoms right after the `conn` is initialized (L461). When you are just checking with `-C`, the atoms will be zero, but that doesn't matter because the matches don't actually have to work.

Airblader replied:
They don't have to work, but if the atoms are zero and I have `for_window[window_type="dialog"] …` in my config, then `i3 -C` will error out telling me that the match is empty. That is, if I use `XCB_NONE` for comparison (which is also `0`).

stapelberg replied:
The obvious question then is: could you use a separate placeholder value instead?

Airblader replied:
Since `xcb_atom_t` is an unsigned integer, I could probably use `-1`. I personally think that it's not a clean approach to deal with uninitialized "constants", but if you want it to be handled that way, I'll change it.

stapelberg replied:
If it’s unsigned, it’d be cleaner to use the MAX value for the corresponding data type, I think. However, in this particular case, it might make more sense to use one of the first 40 atoms because they are hard-coded and their ids cannot change.

stapelberg replied:
With regards to the overall approach: getting the atoms after initializing the X11 connection as @acrisci suggested sounds like the best way to me.

Airblader commented on src/manage.c:
```diff
@@ -374,6 +377,10 @@ void manage_window(xcb_window_t window, xcb_get_window_attributes_cookie_t cooki
+        xcb_reply_contains_atom(type_reply, A__NET_WM_WINDOW_TYPE_MENU) ||
+        xcb_reply_contains_atom(type_reply, A__NET_WM_WINDOW_TYPE_DROPDOWN_MENU) ||
+        xcb_reply_contains_atom(type_reply, A__NET_WM_WINDOW_TYPE_POPUP_MENU) ||
+        xcb_reply_contains_atom(type_reply, A__NET_WM_WINDOW_TYPE_TOOLTIP) ||
```
@acrisci said that this change is unrelated and should be extracted.

My answer is that we set the EWMH hint that we support these atoms, so I think we should handle them here, too. Hence I'd say that it is a related change, especially since it's not too big. I can remove it if wanted, though.

stapelberg replied:
I also think it’s unrelated. “Handling” in this case means setting the window to floating, and I’m not convinced that we should do this for the types you’re adding. If you have a convincing argument, feel free to open another PR, but please keep it separate so that we can revert it easily.

Airblader replied:
I looked at the descriptions of these atoms in the [spec](http://standards.freedesktop.org/wm-spec/latest/ar01s05.html) and it looks to me that all such windows must be floating.

Without this change, i3 tells clients that it supports these atoms, but then opens a dropdown menu as non-floating, which sounds like a bug to me. It probably won't be an issue in practice since all these values are fairly new, so windows will probably define fallbacks to `*_UTILITY` or `*_DIALOG`.

I don't mind a separate SR, but I can't open it until this is merged.

stapelberg replied:
With regards to `_NET_WM_WINDOW_TYPE_MENU` it seems like you’re right, but all the others are (according to the spec) typically set on override-redirect windows, so they will not hit this code path.

stapelberg commented on src/handlers.c:
```diff
@@ -1264,7 +1264,8 @@ static struct property_handler_t property_handlers[] = {
     {0, 128, handle_class_change},
-    {0, UINT_MAX, handle_strut_partial_change}};
+    {0, UINT_MAX, handle_strut_partial_change},
+    {0, UINT32_MAX, handle_window_type}};
```
Why UINT32_MAX instead of UINT_MAX to be consistent with the other lines above?

Airblader commented:
@stapelberg I've changed it to use `xcb_atom_t` now. I went with `UINT32_MAX` as the placeholder, though, because I think this is more descriptive than abusing some unrelated atom (it doesn't make the code very clear). I've also removed handling the new atoms when managing a window.

stapelberg commented on src/commands.c:
```diff
@@ -363,6 +363,31 @@ void cmd_criteria_add(I3_CMD, char *ctype, char *cvalue) {
     }
 
+    if (strcmp(ctype, "window_type") == 0) {
+        if (strcmp(cvalue, "normal") == 0)
```
I think the parser doesn’t canonicalize the values for you here, so please use `strcasecmp` instead of `strcmp`.

stapelberg commented:
Looks good aside from my comment.

Airblader commented:
@stapelberg Done.
